### ISSUE TITLE
net/http/httptest: Server does not check GetCertificate

### ISSUE THREAD

**Issue by mitar** [golang/go#63812]
### What version of Go are you using (`go version`)?

<pre>
go version go1.21.3 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/home/mitar/.cache/go-build'
GOENV='/home/mitar/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/home/mitar/.go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/mitar/.go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/local/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.21.3'
GCCGO='/usr/bin/gccgo'
GOAMD64='v1'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='1'
GOMOD='/dev/null'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build488527063=/tmp/go-build -gno-record-gcc-switches'
</pre></details>

### What did you do?

I am trying to test a `http.Server` instance with httptest's `Server`:

```go
	ts := httptest.NewUnstartedServer(nil)
	ts.EnableHTTP2 = true
	ts.Config = server
	ts.Config.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
	})
	ts.TLS = server.TLSConfig.Clone()
	ts.StartTLS()
```

I want to check that server's TLS configuration works. Server's TLS configuration uses `GetCertificate`

### What did you expect to see?

That once I call `StartTLS` is called, `GetCertificate` is called to obtain the certificate to configure the testing server and testing client.

### What did you see instead?

`StartTLS` ignores and never calls `GetCertificate` but instead configures testing server and testing client to use testing certificate.

### Discussion

I can call something like:

```go

	cert, _ := ts.TLS.GetCertificate(nil) // or some other relevant ClientHelloInfo for 127.0.0.1
	ts.TLS.Certificates = []tls.Certificate{*cert}
```

before calling `StartTLS`. But ideally that should be called by `StartTLS`. This is documented in `tls.Config`:

```
	// Server configurations must set one of Certificates, GetCertificate or
	// GetConfigForClient.
```

Currently only `Certificates` is checked, but `GetCertificate` or `GetConfigForClient` are ignored.

My motivation is to do a test with `golang.org/x/crypto/acme/autocert` serving as manager against [Pebble](https://github.com/letsencrypt/pebble) to test that `http.Server`'s configuration and everything works correctly.

**Comment by rittneje** [comment]
The source of the issue is here:
https://github.com/golang/go/blob/883f062fc0a097bf561030ad453fd3e300896975/src/net/http/httptest/server.go#L167-L169

I guess it could also check if `GetCertificate != nil`. But then I don't know what the [`Certificate`](https://pkg.go.dev/net/http/httptest#Server.Certificate) method is supposed to return in that case.

I'll also note that this behavior of `httptest.Server` seems to be completely undocumented.

**Comment by mitar** [comment]
I thnk `GetCertificate` should be called with `127.0.0.1` (or more precisely `s.Listener.Addr().String()`) as a parameter. That is what it gets set in `s.URL`.

**Comment by rittneje** [comment]
`GetCertificate` takes an entire [`ClientHelloInfo`](https://pkg.go.dev/crypto/tls#ClientHelloInfo), not just a hostname.

**Comment by mitar** [comment]
@rittneje, yes, I know. What I mean that it should be a dummy `ClientHelloInfo` with `s.Listener.Addr()`.

**Comment by seankhliao** [comment]
I think this is out of scope, if you already have a complete http.Server, you should be testing that directly, rather than some mutated version of it with httptest.Server.

**Comment by mitar** [comment]
Testing server does a lot of useful plumbing real server does not: like exposing a client which terminates all connections when test finishes.
